# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/forms/nsTextControlFrame.cpp
# Commit: 2187935f6d78
# Full Hash: 2187935f6d78c22ceeb7e5e92e919aeb42820b95
# Author: David Shin <dshin@mozilla.com>
# Date: 2023-08-16 03:43:43
# Regressor Bug: 1846982
# File Overlap Count: 1
# Description:
#   Bug 1846982: Baseline calculation for textarea. r=emilio
#   
#   Tests are added as tentative, because the behaviour when there's no content,
#   or when there's placeholder text, seem very differnt for each browser.
#   
# ==============================================================================

diff -r 2d475d0406cd -r 2187935f6d78 layout/forms/nsTextControlFrame.cpp
--- a/layout/forms/nsTextControlFrame.cpp	Tue Aug 15 13:26:25 2023 +0000
+++ b/layout/forms/nsTextControlFrame.cpp	Tue Aug 15 13:30:17 2023 +0000
@@ -1351,3 +1351,26 @@
     nsIContent* aChild, nsIContent* aPreviousSibling) {
   mFrame.ClearCachedValue();
 }
+
+Maybe<nscoord> nsTextControlFrame::GetNaturalBaselineBOffset(
+    mozilla::WritingMode aWM, BaselineSharingGroup aBaselineGroup,
+    BaselineExportContext aExportContext) const {
+  if (!IsSingleLineTextControl()) {
+    if (StyleDisplay()->IsContainLayout()) {
+      return Nothing{};
+    }
+
+    if (aBaselineGroup == BaselineSharingGroup::First) {
+      return Some(std::clamp(mFirstBaseline, 0, BSize(aWM)));
+    }
+    // This isn't great, but the content of the root NAC isn't guaranteed
+    // to be loaded, so the best we can do is the edge of the border-box.
+    if (aWM.IsCentralBaseline()) {
+      return Some(BSize(aWM) / 2);
+    }
+    return Some(0);
+  }
+  NS_ASSERTION(!IsSubtreeDirty(), "frame must not be dirty");
+  return GetSingleLineTextControlBaseline(this, mFirstBaseline, aWM,
+                                          aBaselineGroup);
+}